people-and-planet-ai/land-cover-classification/trainer.py

```diff
@@ -0,0 +1,227 @@
+            num_parallel_calls=tf.data.AUTOTUNE,
+        )
+        .map(preprocess, num_parallel_calls=tf.data.AUTOTUNE)
+        # Finally we cache the current batch and prefetch the next one.
+        .cache()
+        .prefetch(tf.data.AUTOTUNE)
+        # For more information on how to optimize your tf.data.Dataset, see:
+        #   https://www.tensorflow.org/guide/data_performance
+    )
+
+
+def new_model(training_dataset: tf.data.Dataset) -> tf.keras.Model:
+    """Creates a new Fully Convolutional Network (FCN) model.
+
+    Args:
+        training_dataset: The dataset to use for the normalization layer.
+
+    Returns: A Fully Convolutional Network model.
+    """
+
+    # Adapt the Normalization layer with the training dataset.
+    normalization = tf.keras.layers.Normalization()
+    normalization.adapt(
+        training_dataset.map(
+            lambda inputs, _: tf.stack([inputs[name] for name in INPUT_BANDS], axis=-1)
+        )
+    )
+
+    # Define the Fully Convolutional Network.
+    layers = [
+        tf.keras.Input(shape=(None, None, len(INPUT_BANDS))),
+        normalization,
+        tf.keras.layers.Conv2D(filters=32, kernel_size=5, activation="relu"),
+        tf.keras.layers.Conv2DTranspose(filters=16, kernel_size=5, activation="relu"),
+        tf.keras.layers.Dense(NUM_CLASSIFICATIONS, activation="softmax"),
+    ]
+    fcn_model = tf.keras.Sequential(layers, name="FullyConvolutionalNetwork")
+
+    # Define the input dictionary layers.
+    input_layers = {
+        name: tf.keras.Input(shape=(None, None, 1), name=name) for name in INPUT_BANDS
+    }
+
+    # Model wrapper that takes an input dictionary and feeds it to the FCN.
+    inputs = tf.keras.layers.concatenate(input_layers.values())
+    model = tf.keras.Model(input_layers, fcn_model(inputs))
+
+    model.compile(
+        optimizer="adam",
+        loss="categorical_crossentropy",
+        metrics=["accuracy"],
+    )
+    return model
+
+
+def run(
+    training_data: str,
+    validation_data: str,
+    model_path: str,
+    patch_size: int,
+    epochs: int,
+    batch_size: int = 256,
+) -> None:
+    """Creates, trains and saves a new model.
+
+    Args:
+        training_data: File pattern for the training data files.
+        validation_data: File pattern for the validation data files.
+        model_path: Path to save the model to.
+        patch_size: Patch size of the training and validation datasets.
+        epochs: Number of times to go through the training dataset.
+        batch_size: Number of examples for the model to look at the same time.
+    """
+
+    # Read the training and validation datasets.
+    training_dataset = read_dataset(training_data, patch_size, batch_size)
+    validation_dataset = read_dataset(validation_data, patch_size, batch_size)
+
+    # Create, train and save the model.
+    model = new_model(training_dataset)
+    model.fit(
+        training_dataset.shuffle(10),
```

Comment:
I think you would want to shuffle before you batch. Otherwise you are just shuffling within the batches themselves.

Reply:
If we shuffle before caching, wouldn't that make all the batches the same anyways? The data points should already be in a random order due to Dataflow and the interleaving, right?

Reply:
never mind, shuffling after the batching doesn't shuffle within the individual batches. It just shuffles the order of all the batches. Either way, what is the purpose of doing that, since as you mention the data is already in a random order?

Reply:
I've looked into this a few years ago, here was a good resource for shuffling and randomness: https://www.moderndescartes.com/essays/shuffle_viz/

Reply:
I guess the idea was to feed the batches in a "random" order so each epoch would be different, but it might not make any difference.

Should I just remove the shuffle? Or do you think it still adds value? Or should we shuffle before batching?

Reply:
I removed the shuffle for now, but please let me know if you think we should still do it in any other way.

Reply:
I think we should visualize the data shuffling to see the optimal shuffling technique. I'll run some simulations.